lib/class_relationships.rb

```diff
@@ -0,0 +1,67 @@
+# Think of these toy objects as the singletons
+ToyObject = BasicObject.new
+
+def ToyObject.to_s
```

Comment:
I've definitely used `class << self` in Ruby classes before to reopen the singleton class and define class-level methods, like this:
```ruby
class Foo
  class << self 
    def my_singleton_method
    end
  end
end
```
Is it possible to make this work with our `BasicObject` classes? 😄  Or is there another approach altogether that I have not yet seen?!

Reply:
> Is it possible to make this work with our `BasicObject` classes?

Yes! Note that `self` isn’t special syntax in the `class << self` you wrote, it’s just a different way of referring to `Foo`. So you can just as well write `class << Foo`, or indeed `class << any_object_at_all`, to open up a singleton class for method definitions:

```ruby
>> greeting = 'Hello, world'
=> "Hello, world"
>> class << greeting
     def shout
       upcase + '!!!'
     end
   end
=> :shout
>> greeting.shout
=> "HELLO, WORLD!!!"
>> 'bye'.shout
in `<main>': undefined method `shout' for "bye":String (NoMethodError)
```

Reply:
Actually, I’ve realised I can grasp another opportunity for ✨ barely relevant Ruby trivia ✨ here. Lucky you!

We’re familiar with `self` as always referring to the “current object” within the context where it appears, e.g. inside your `class Foo` example above, the use of `self` refers to `Foo` because that’s the class currently being defined. Arguably the most important role of `self` is to act as the default receiver of messages: if we say `a.b` then that’s explicitly sending the message `:b` to the object `a`, but if we say simply `b` (and `b` isn’t a local variable) then it implicitly means sending the message `:b` to `self`, i.e. `self.b`.

But there’s another piece of invisible context that changes like `self`, which is: what’s the current _module_? Or, by analogy with `self`, what’s the default destination of method definitions? If we say `def a.b` then we’re explicitly defining the method `#b` in `a`’s singleton class, but if we say simply `def b`, in which module are we implicitly placing the definition of `#b`?

This invisible context is hard to talk about because it doesn’t have a concrete name like `self` does, so I guess let’s just call it the “current module”. It’s also hard to even notice that the distinction exists, because the most common case is something like…

```
>> class Foo
     puts "self is #{self}"
     def bar
       # …
     end
   end
self is Foo
=> :bar
>> Foo.instance_methods(false) # exclude inherited methods
=> [:bar]
```

…where the “current object” and “current module” inside `class Foo … end` are the same object, i.e. `Foo`. This is convenient because it makes vanilla class definitions easy to write and understand.

But the two concepts are actually independent, which is most evident in the distinction between [`BasicObject#instance_eval`](https://rubydoc.info/stdlib/core/BasicObject:instance_eval) and [`Module#module_eval`](https://rubydoc.info/stdlib/core/Module:module_eval). Both of these methods evaluate a block, during which `self` will be set to the object you called the method on (i.e. the receiver):

```
>> Foo = Class.new
=> Foo
>> Foo.instance_eval { puts "self is #{self}" }
self is Foo
=> nil
>> Foo.module_eval { puts "self is #{self}" }
self is Foo
=> nil
```

At first glance they appear interchangeable because they do the same thing with respect to `self`, but actually they differ in what value they use as the current module during the block: for `#instance_eval` it’s the receiver’s singleton class, whereas for `#module_eval` it’s the receiver itself (which is always a module, because `#module_eval` is only implemented for modules). So an unqualified `def` puts a method definition in a different module in each case:

```
>> Foo.instance_eval { def a; :a; end }
=> :a
>> Foo.module_eval { def b; :b; end }
=> :b

>> Foo.a
=> :a
>> Foo.new.a
in `<main>': undefined method `a' for #<Foo> (NoMethodError)

>> Foo.new.b
=> :b
>> Foo.b
in `<main>': undefined method `b' for Foo:Class (NoMethodError)

>> Foo.singleton_class.instance_methods(false)
=> [:a]
>> Foo.instance_methods(false)
=> [:b]
```

I suppose the tldr is “`#instance_eval` is yet another way to define methods in an object’s singleton class”. But there’s some deeper knowledge here about the fact that Ruby is tracking this “current module” context at all times and using it to understand which module we’re talking about when we say `def`, and that’s not necessarily (although usually is!) the same object we get when we ask for `self`.